### DISCUSSION TITLE
jUnit 4.x in dropwizard 3?

### DISCUSSION THREAD

**Discussion by rhowe** [dropwizard/dropwizard#5382] [Development]
I raised this PR a while back as a suggestion: https://github.com/dropwizard/dropwizard/pull/5138

Is jUnit 4 worth keeping around in dropwizard 3.x? There was some discussion on the PR, but figured I'd open a thread here for some chatter.

**Comment by zUniQueX** [comment]
I'd vote to drop support for it. Java 11 was released after jUnit 5.x and we will use a Java 11 baseline in 3.x. jUnit 5 is mature enough and we could now take the chance to drop some old dependencies.

**Reply by joschi** [reply]
For people who absolutely do not want to miss JUnit 4.x support we could extract that functionality in its own Dropwizard module similar to https://github.com/dropwizard/dropwizard-jdbi and maintain it on a best-effort base or find a new maintainer.

**Reply by rhowe** [reply]
If we were to do that, should we lazy-initialise the repo (i.e. create it if someone asks for it), or is it worth doing the work up-front do you think?

**Reply by pstackle** [reply]
My vote would be to keep it. While JUnit 5 is definitely mature at this point, that doesn't mean that all other 3rd party test support has made the jump from using JUnit 4 Rules to JUnit 5 Extensions. That could introduce a pain point or at least additional friction for teams trying to upgrade to Dropwizard v3. Most (all?) of the logic lives in the `DropwizardTestSupport` and `DAOTest` classes anyway, so the JUnit 4 support, which is already written, _should_ be minimal effort to maintain.

Since we are already changing the minimum Java version and the Jetty version, and I don't think we save much with the removal at this point, my preference would be to keep the JUnit 4 rules for the Dropwizard v3 and v4 releases.

**Reply by joschi** [reply]
I've extracted the JUnit 4.x classes into a separate module:
https://github.com/dropwizard/dropwizard-testing-junit4

Except for having to add an additional dependency, it's a drop-in replacement for the JUnit 4.x functionality in Dropwizard 2.x.

**Reply by rhowe** [reply]
I think that's cleaner than people adding exclude rules for junit4 if they are pulling in dropwizard-testing and don't want 4.x in their classpath
